File: 7a/WordDictionary.java
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;

public class WordDictionary {
  private int max = 11;
  private int size = 0;
  private int cur = 0;
  private String[] table = new String[max];
  private ArrayList<String> next = new ArrayList<>();

  public void addWord(String word) {
    if (word.trim().length() == 0 || (word.length() == 1 && !word.equals("a") && !word.equals("i") &&
        !word.equals("A") && !word.equals("I"))) {
      return;
    }

    if (isInDictionary(word)) {
      return;

    }

    table[rehash(word, 0, table)] = word;

    size += 1;

    if ((double) size / max > .5) {
      resize();
    }
  }

  public int size() {
    return size;
  }

  private void resize() {
    max = max * 2;
    String[] newTable = new String[max];

    for (String word : table) {
      if (word != null) {
        newTable[rehash(word, 0, newTable)] = word;
      }
    }

    table = newTable;
  }

  private int rehash(String word, int i, String[] table) {
    int pos = (Math.abs(word.toLowerCase().hashCode()) + (int) Math.pow(i, 2)) % max;

    if (table[pos] == null) {
      return pos;
    } else {
      return rehash(word, i + 1, table);
    }
  }

  public String nextWord() {
    for (int i = 0; i < max; i++) {
      if (!next.contains(table[i]) && table[i] != null) {
        next.add(table[i]);
      }
    }
    if (cur >= size)
      return null;

    String ans = next.get(cur);

    cur++;

    return ans;
  }

  public void reset() {
    cur = 0;
  }

  public void write(String filename) {
    try (
        FileWriter writer = new FileWriter(filename);
        BufferedWriter csv = new BufferedWriter(writer)) {
      for (int i = 0; i < max; i++) {
        if (table[i] != null) {
          csv.append(table[i]).append("\n");
        }
      }
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  private boolean isInDictionary(String word) {
    int pos = Math.abs(word.toLowerCase().hashCode()) % max;

    int i = 1;
    while (table[pos] != null) {
      if (table[pos].equalsIgnoreCase(word))
        return true;

      pos = ((Math.abs(word.toLowerCase().hashCode()) + (int) Math.pow(i, 2))) % max;
      i++;
    }

    return false;
  }

  public static void main(String[] args) {
    WordDictionary wd = new WordDictionary();
    // System.out.println("Description \t\t\t Expected \t\t Output");
    // System.out.println("----------- \t\t\t -------- \t\t ------");

    // print("empty size", "0", "size is " + wd.size());

    // wd.addWord("");
    // print("add empty String", "0", "size is " + wd.size());

    // print("empty String nextWord", "", wd.nextWord());

    // print("size after adding empty String", "0",
    // Integer.toString(wd.size()));

    // wd.addWord("test");
    // print("add test and nextWord", "test", wd.nextWord());

    // print("nextWord when nextNode is null (at end)", "", wd.nextWord());

    // wd.addWord("TeST");
    // wd.reset();
    // print("add TeST and reset and nextWord", "test", wd.nextWord());

    // wd.reset();
    // print("reset with one item and nextWord", "test", wd.nextWord());

    // wd.addWord("a");
    // print("add a then size", "2", Integer.toString(wd.size()));
    // wd.reset();
    // String temp1 = wd.nextWord();
    // print("reset and nextWord", "nextWord is a or test",
    // (temp1 == "a" || temp1 == "test")
    // ? "nextWord is a or test"
    // : "nextWord is not a or test");

    // String temp2 = wd.nextWord();
    // print("nextWord again", "nextWord is a or test",
    // (temp2 == "a" || temp2 == "test")
    // ? "nextWord is a or test"
    // : "nextWord is not a or test");

    // wd.addWord("A");
    // wd.reset();
    // String temp3 = wd.nextWord();
    // print("add A then nextWord", "nextWord is a or test",
    // (temp3 == "a" || temp3 == "test")
    // ? "nextWord is a or test"
    // : "nextWord is not a or test");

    // String temp4 = wd.nextWord();
    // print("nextWord again", "nextWord is a or test",
    // (temp4 == "a" || temp4 == "test")
    // ? "nextWord is a or test"
    // : "nextWord is not a or test");

    // wd.addWord("m");
    // print("add m then IsInDictionary m", "false",
    // wd.isInDictionary("m") ? "true" : "false");

    // print("IsInDictionary test", "true",
    // wd.isInDictionary("test") ? "true" : "false");

    // print("IsInDictionary TeST", "true",
    // wd.isInDictionary("TeST") ? "true" : "false");

    // print("IsInDictionary i aint here", "false",
    // wd.isInDictionary("i aint here") ? "true" : "false");

    // for (int i = 10; i < 1000; i++) {
    // wd.addWord(Integer.toString(i));
    // }

    // print("size after adding 990 items", "992", wd.size());
    for (int i = 0; i < 50010; i++) {
      wd.addWord(Integer.toString(i));
    }
    System.out.println(wd.size());

    wd.write("WordDictionary.csv");
  }

  private static <D, E, O> void print(D description, E expected, O output) {
    System.out.println(description + " \t\t\t " + "expected " + expected + "\t\t" + "outputted " + output);
  }
}

File: WordDictionary/WordDictionary.cpp
#include "WordDictionary.h"
#include <cassert>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

WordDictionary::WordDictionary() {
  head = NULL;
  nextNode = NULL;
  size = 0;
}

Node* WordDictionary::getHead() { return head; }

bool WordDictionary::compareStringsCaseInsensitive(string s1, string s2) {
  if (s1.length() != s2.length()) {
    return false;
  }

  for (int i = 0; i < s1.length(); i++) {
    if (tolower(s1.at(i)) != tolower(s2.at(i))) {
      return false;
    }
  }

  return true;
}

bool WordDictionary::isInList(Node* node, string str) {
  Node* cur = node;
  while (cur != NULL) {
    if (compareStringsCaseInsensitive(cur->value, str)) {
      return true;
    }
    cur = cur->next;
  }
  return false;
}

void WordDictionary::addWord(string word) {
  if (word.length() == 0 || (word.length() == 1 && word != "a" && word != "i" &&
                             word != "A" && word != "I")) {
    return;
  }

  if (isInDictionary(head, word)) {
    return;
  }

  Node* node = new Node();
  node->value = word;

  if (head == NULL) {
    head = node;
  } else {
    node->next = head;
    head = node;
  }

  if (nextNode == NULL)
    nextNode = node;

  size++;
}

int WordDictionary::length() { return size; }

string WordDictionary::nextWord() {
  if (nextNode == NULL) {
    return "";
  }

  string res = nextNode->value;
  nextNode = nextNode->next;

  return res;
}

bool WordDictionary::isInDictionary(Node* cur, string target) {
  if (cur == NULL) {
    return false;
  }
  if (compareStringsCaseInsensitive(cur->value, target)) {
    return true;
  } 

  return isInDictionary(cur->next, target);
}

void WordDictionary::reset() { nextNode = head; }

void WordDictionary::print() {
  Node* cur = head;
  while (cur != NULL) {
    cout << cur->value << endl;
    cur = cur->next;
  }
}

void WordDictionary::write(string fileName) {
  ofstream file;
  file.open(fileName);

  Node* cur = head;
  while (cur != NULL) {
    file << cur->value << endl;
    cur = cur->next;
  }
}

void WordDictionary::read(string fileName) {
  ifstream file;
  file.open(fileName, ios::in);
  string temp;

  vector<string> stack;
  while (getline(file, temp)) {
    stack.push_back(temp);
  }

  for (int i = stack.size() - 1; i >= 0; i--) {
    this->addWord(stack.at(i));
  }
}

void WordDictionary::prettyPrintTest(string description, string expected,
                                     string actual) {
  const char separator = ' ';
  cout << description << setw(30) << setfill(separator) << expected << setw(20)
       << setfill(separator) << actual << setw(20) << setfill(separator)
       << endl;
}

bool WordDictionary::testEverythingCopied() {
  for (int i = 10; i < 1000; i++) {
    if (!this->isInDictionary(head, to_string(i))) {
      return false;
    }
  }

  if (!this->isInDictionary(head, "a") || !this->isInDictionary(head, "test")) {
    return false;
  }

  return true;
}

void WordDictionary::test() {
  WordDictionary wd;
  prettyPrintTest("Description", "Expected", "Actual");

  prettyPrintTest("empty size", "0", to_string(wd.length()));

  wd.addWord("");
  prettyPrintTest("add empty string", "", wd.nextWord());

  prettyPrintTest("size after adding empty string", "0",
                  to_string(wd.length()));

  wd.addWord("test");
  prettyPrintTest("add test and nextWord", "test", wd.nextWord());

  prettyPrintTest("nextWord when nextNode is null (at end)", "", wd.nextWord());

  wd.addWord("TeST");
  wd.reset();
  prettyPrintTest("add TeST and reset and nextWord", "test", wd.nextWord());

  wd.reset();
  prettyPrintTest("reset with one item and nextWord", "test", wd.nextWord());

  wd.addWord("a");
  prettyPrintTest("add a then size", "2", to_string(wd.length()));
  wd.reset();
  string temp1 = wd.nextWord();
  prettyPrintTest("reset and nextWord", "nextWord is a or test",
                  (temp1 == "a" || temp1 == "test")
                      ? "nextWord is a or test"
                      : "nextWord is not a or test");

  string temp2 = wd.nextWord();
  prettyPrintTest("nextWord again", "nextWord is a or test",
                  (temp2 == "a" || temp2 == "test")
                      ? "nextWord is a or test"
                      : "nextWord is not a or test");

  wd.addWord("A");
  wd.reset();
  string temp3 = wd.nextWord();
  prettyPrintTest("add A then nextWord", "nextWord is a or test",
                  (temp3 == "a" || temp3 == "test")
                      ? "nextWord is a or test"
                      : "nextWord is not a or test");

  string temp4 = wd.nextWord();
  prettyPrintTest("nextWord again", "nextWord is a or test",
                  (temp4 == "a" || temp4 == "test")
                      ? "nextWord is a or test"
                      : "nextWord is not a or test");

  wd.addWord("m");
  prettyPrintTest("add m then isInDictionary m", "false",
                  wd.isInDictionary(wd.getHead(), "m") ? "true" : "false");

  prettyPrintTest("isInDictionary test", "true",
                  wd.isInDictionary(wd.getHead(), "test") ? "true" : "false");

  prettyPrintTest("isInDictionary TeST", "true",
                  wd.isInDictionary(wd.getHead(), "TeST") ? "true" : "false");

  prettyPrintTest("isInDictionary i aint here", "false",
                  wd.isInDictionary(wd.getHead(), "i aint here") ? "true" : "false");

  for (int i = 10; i < 1000; i++) {
    wd.addWord(to_string(i));
  }

  prettyPrintTest("size after adding 990 items", "992", to_string(wd.length()));

  wd.write("input.txt");

  WordDictionary readWd;
  readWd.read("input.txt");
  readWd.reset();
  prettyPrintTest("write and read into new object", "992",
                  to_string(wd.length()));

  prettyPrintTest("everything is copied", "true",
                  readWd.testEverythingCopied() ? "true" : "false");
}